### PR TITLE
Add Occluded events based on foreground/background on iOS and MemoryWarning on iOS/Android

### PR DESCRIPTION
**Pull request by mockersf**
Adopted from #2144, and updated.

From https://github.com/rust-windowing/winit/pull/2144#issuecomment-1129990322 I'm not confident on adding events `Foreground` and `Background` to Android, they seem to be equivalent to `Suspend` / `Resume` where iOS split that in two steps.

- [ ] Tested on all platforms changed
- [ ] Added an entry to `CHANGELOG.md` if knowledge of this change could be valuable to users
- [ ] Updated documentation to reflect any user-facing changes, including notes of platform-specific behavior
- [ ] Created or updated an example program if it would help users understand this functionality
- [ ] Updated [feature matrix](https://github.com/rust-windowing/winit/blob/master/FEATURES.md), if new features were added or implemented

### CODE REVIEW

**Comment by madsmtm** [comment]
That's really nice, thank you!

From my reading, to be maximally cross-platform compatible, it sounds like we should send an artificial `Background` event just before `Suspend` on Android, and an artificial `Foreground` just after `Resumed` (on all platforms, actually)?

**Comment by madsmtm** [comment]
Hmm, actually it sounds like `applicationDidEnterBackground:` matches the semantics of the `Suspend` event on Android better than `applicationWillResignActive`?

**Comment by madsmtm** [comment]
@rib: Is there an event on Android that requests the application to pause things (e.g. pause the game) and go into a low-power state, without necessarily suspending _everything_?

**Comment by mockersf** [comment]
> From my reading, to be maximally cross-platform compatible, it sounds like we should send an artificial `Background` event just before `Suspend` on Android, and an artificial `Foreground` just after `Resumed` (on all platforms, actually)?

Is it OK if I add those artificial events on other platforms, with a TODO to move them to the correct events on Android for now?

**Comment by daxpedda** [comment]
Web should probably use these for the `pagehide` and `pageshow` events, currently it uses `Occluded` instead.
Should we also emit `Occluded` with `Foreground` and `Background`?
Should this be folded into the new `WindowState` (https://github.com/rust-windowing/winit/pull/2929)?

**Comment by kchibisov** [comment]
If the whole intent is to stop rendering based on foreground/background `Occluded` sounds like a better choice.

**Comment by daxpedda** [comment]
My suggestion was to use **both**, `Occluded` and `Foreground`/`Background`.

Maybe I'm misunderstanding `Foreground`/`Background`, my assumption was that if something is in the background it's not visible, ergo occluded.

**Comment by kchibisov** [comment]
I mean, what's the reason to add them, if `Occluded` could indicate `Foreground`/`Background`?

**Comment by kchibisov** [comment]
@daxpedda it's simply a hint saying that you could stop rendering.

**Comment by daxpedda** [comment]
> I mean, what's the reason to add them, if `Occluded` could indicate `Foreground`/`Background`?

> @daxpedda it's simply a hint saying that you could stop rendering.

I agree, that's why I am proposing to send both, because `Occluded` should not indicate anything else then to stop rendering.

**Comment by kchibisov** [comment]
@daxpedda I'm just arguing against adding special events for `ios` where we already can use existing `Occluded` event.

**Comment by mockersf** [comment]
`Occluded` is a `WindowEvent`, while `Foreground`/`Background` are application events. When dealing with them, you're not bound to any specific window, and putting them as `WindowEvent`would (in my opinion) break some of their meaning.

**Comment by kchibisov** [comment]
>Occluded is a WindowEvent, while Foreground/Background are application events. When dealing with them, you're not bound to any specific window, and putting them as WindowEventwould (in my opinion) break some of their meaning.

>That said, that limit on iOS at least is quite blurry as it's not possible to have more than one window, so it may be ok to mix those there...

I mean, as long as window is not destroyed it's fine to use `Occluded` for that.

**Comment by mockersf** [comment]
Moved to `WindowEvent::Occluded`

File: src/platform_impl/ios/view.rs [hunk 1]
```diff
@@ -525,10 +525,46 @@ declare_class!(
...
+                        &*ptr
+                    };
+                    events.push(EventWrapper::StaticEvent(Event::WindowEvent {
+                        window_id: RootWindowId(window.id()),
+                        event: WindowEvent::Occluded(true),
+                    }));
+                }
+            }
+            let mtm = MainThreadMarker::new().unwrap();
+            app_state::handle_nonuser_events(mtm, events);
```
Comment (kchibisov) [inline src/platform_impl/ios/view.rs:566]: Could you extract the code into common function? The only difference is `true/false` to `Occluded`, really.

File: src/event.rs [hunk 2]
```diff
@@ -225,6 +225,26 @@ pub enum Event<T: 'static> {
...
+    ///
+    /// ## iOS
+    ///
+    /// On iOS, the `MemoryWarning` event is emitted in response to an [`applicationDidReceiveMemoryWarning`]
+    /// callback. The application must free as much memory as possible or risk being terminated, see
+    /// [how to respond to memory warnings].
+    ///
+    /// [`applicationDidReceiveMemoryWarning`]: https://developer.apple.com/documentation/uikit/uiapplicationdelegate/1623063-applicationdidreceivememorywarni
+    /// [how to respond to memory warnings]: https://developer.apple.com/documentation/uikit/app_and_environment/managing_your_app_s_life_cycle/responding_to_memory_warnings
+    MemoryWarning,
```
Comment (kchibisov) [inline src/event.rs:247]: Should do a proper `## Platform-specific` section with a separate android/ios entries in it, saying that the rest(wayland, macOS) is not supported.

File: src/event.rs [hunk 3]
```diff
@@ -533,8 +554,19 @@ pub enum WindowEvent {
     ///
     /// Platform-specific behavior:
     ///
+    /// ## iOS
```
Comment (kchibisov) [inline src/event.rs:557]: Could you change the `Platform-specific behavior:` to `## Platform specific` and use `### iOS` while at it?

**Review by kchibisov — CHANGES_REQUESTED** [review]

File: src/event.rs (hunk 3 above)
Comment (mockersf) [reply src/event.rs]: used `## Platform-specific` to be in line with the others

**Review by kchibisov — APPROVED** [review]

**Pull request merged**